src/blindpsbt.cpp

```diff
@@ -0,0 +1,430 @@
+}
+
+uint256 GenerateRangeproofECDHKey(CPubKey& ephemeral_pubkey, const CPubKey blinding_pubkey)
+{
+    // Generate ephemeral key for ECDH nonce generation
+    CKey ephemeral_key;
+    ephemeral_key.MakeNewKey(true);
+    ephemeral_pubkey = ephemeral_key.GetPubKey();
+    assert(ephemeral_pubkey.size() == CConfidentialNonce::nCommittedSize);
+    // Generate nonce
+    uint256 nonce = ephemeral_key.ECDH(blinding_pubkey);
+    CSHA256().Write(nonce.begin(), 32).Finalize(nonce.begin());
+    return nonce;
+}
+
+bool CreateValueRangeProof(std::vector<unsigned char>& rangeproof, const uint256& value_blinder, const uint256& nonce, const CAmount amount, const CScript& scriptPubKey, const secp256k1_pedersen_commitment& value_commit, const secp256k1_generator& gen, const CAsset& asset, const uint256& asset_blinder)
+{
+    // Prep range proof
+    size_t rangeproof_len = 5134;
+    rangeproof.resize(rangeproof_len);
+
+    // Compose sidechannel message to convey asset info (ID and asset blinds)
+    unsigned char asset_message[SIDECHANNEL_MSG_SIZE];
+    memcpy(asset_message, asset.begin(), 32);
+    memcpy(asset_message + 32, asset_blinder.begin(), 32);
+
+    // Sign rangeproof
+    int ct_exponent = (int)gArgs.GetArg("-ct_exponent", 0);
+    int ct_bits = (int)gArgs.GetArg("-ct_bits", 52);
+    // If min_value is 0, scriptPubKey must be unspendable
+    uint64_t min_value = scriptPubKey.IsUnspendable() ? 0 : 1;
+    int res = secp256k1_rangeproof_sign(secp256k1_blind_context, rangeproof.data(), &rangeproof_len, min_value, &value_commit, value_blinder.begin(), nonce.begin(), ct_exponent, ct_bits, amount, asset_message, sizeof(asset_message), scriptPubKey.size() ? &scriptPubKey.front() : NULL, scriptPubKey.size(), &gen);
+    rangeproof.resize(rangeproof_len);
+    return (res == 1);
+}
+
+void CreateAssetCommitment(CConfidentialAsset& conf_asset, secp256k1_generator& asset_gen, const CAsset& asset, const uint256& asset_blinder)
+{
+    conf_asset.vchCommitment.resize(CConfidentialAsset::nCommittedSize);
+    int ret = secp256k1_generator_generate_blinded(secp256k1_blind_context, &asset_gen, asset.begin(), asset_blinder.begin());
+    assert(ret == 1);
+    ret = secp256k1_generator_serialize(secp256k1_blind_context, conf_asset.vchCommitment.data(), &asset_gen);
+    assert(ret != 0);
+}
+
+void CreateValueCommitment(CConfidentialValue& conf_value, secp256k1_pedersen_commitment& value_commit, const uint256& value_blinder, const secp256k1_generator& asset_gen, const CAmount amount)
+{
+    int ret;
+    conf_value.vchCommitment.resize(CConfidentialValue::nCommittedSize);
+    ret = secp256k1_pedersen_commit(secp256k1_blind_context, &value_commit, value_blinder.begin(), amount, &asset_gen);
+    assert(ret != 0);
+    secp256k1_pedersen_commitment_serialize(secp256k1_blind_context, conf_value.vchCommitment.data(), &value_commit);
+    assert(conf_value.IsValid());
+}
+
+// Subtract b from a in place
+bool SubtractScalars(uint256& a, const uint256& b)
```

Comment:
Probably best to do this in a followup PR, but Sanket found a simpler way to do all this arithmetic by calling `secp256k1_pedersen_blind_generator_blind_sum` in secp-zkp.